matrix.py

```diff
@@ -0,0 +1,156 @@
+"""Matrix exercise, Daniel Prince - unity training."""
+
+
+class Matrix:
+    """Perform all sorts of matrix operations."""
+
+    def __init__(self, data):
+        self.data = data
+        # Check if n x n
+        for tup in self.data:
+            if len(self.data) != len(tup):
+                raise ValueError("not a valid matrix. (not nXn)")
+
+    def __str__(self):
+        return str(self.data)
+
+    def __repr__(self):
+        return '{self.__class__.__name__}({self.data})'.format(self=self)
+
+    def __iter__(self):
+        return iter(self.data)
+
+    def __hash__(self):
+        return hash(self.data)
+
+    def __eq__(self, other):
+        return self.data == other.data
+
+    def __ne__(self, other):
+        return self.data != other.data
+
+    def __add__(self, other):
+        """Adds 2 matrices."""
+
+        if isinstance(other, Matrix):
+            if len(self.data) == len(other.data):
+                post_add_list = [self.data[row_i][column_i]
+                                 + other.data[row_i][column_i]
+                                 for row_i in range(len(self.data))
+                                 for column_i in range(len(other.data))]
```

Comment:
Instead of iterating through the array's length and accessing every element every time, iterate through the array itself. 
For example if you wish to print an array you could do:
`for i in range(len(array)): print(array[i])`

Or you could simply do:
`for val in array: print(val)`

Reply:
I thought of that originally, but the list I need to iterate over is made up of smaller lists.
so looping over the elements themselves wouldn't give me the number I need, it would give me a list. 
this can still be done but I was told to use list comprehension and I don't know of a way of doing it using list comprehension.